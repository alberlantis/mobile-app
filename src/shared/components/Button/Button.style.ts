import { StyleSheet } from "react-native";

import colors from "src/theme/colors";

export default StyleSheet.create({
  container: {
    justifyContent: "center",
    alignItems: "center",
  },
  contentContainer: {
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
  },
  defaultText: {
    color: colors.WHITE,
    fontSize: 16,
    fontWeight: "medium",
  },
});
