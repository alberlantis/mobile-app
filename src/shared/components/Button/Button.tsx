import React, { useCallback, useState } from "react";
import {
  ColorValue,
  DimensionValue,
  GestureResponderEvent,
  Pressable,
  StyleSheet,
  Text,
  View,
  LayoutChangeEvent,
} from "react-native";
import Svg, {
  Defs,
  ClipPath,
  LinearGradient,
  Stop,
  Rect,
} from "react-native-svg";

import type { IconProps, IconType } from "../Icon";
import s from "./Button.style";
import colors from "src/theme/colors";

type ButtonSize = "large" | "extra-large" | "regular" | "auto";
type ButtonTheme = "primary" | "secondary" | "disabled";
type NumberProp = string | number;
type ButtonDimensions = {
  width?: NumberProp & DimensionValue;
  height: NumberProp & DimensionValue;
};

export interface IButtonProps {
  size?: ButtonSize;
  theme: ButtonTheme;
  text: string;
  marginTop?: number;
  marginBottom?: number;
  marginLeft?: number;
  marginRight?: number;
  buttonIcon?(): React.ComponentElement<IconProps<IconType>, any>;
  onPress(event?: GestureResponderEvent): void;
}

const getColors = (theme: ButtonTheme): ColorValue[] => {
  switch (theme) {
    case "secondary":
      return [colors.BLUE_PRIMARY, colors.BLUE_PRIMARY];
    case "disabled":
      return [colors.GRAY_BOLD, colors.GRAY_BOLD];
    case "primary":
    default:
      return [colors.ORANGE_PRIMARY_DARK, colors.ORANGE_PRIMARY_LIGHT];
  }
};
const getSize = (size?: ButtonSize): ButtonDimensions => {
  switch (size) {
    case "regular": {
      return {
        width: 130,
        height: 40,
      };
    }
    case "extra-large": {
      return {
        width: "100%",
        height: 50,
      };
    }
    case "large":
      return {
        width: 300,
        height: 50,
      };
    default:
      return {
        height: 40,
      };
  }
};

const Button: React.FC<IButtonProps> = ({
  size,
  theme,
  text,
  marginTop,
  marginBottom,
  marginLeft,
  marginRight,
  buttonIcon,
  onPress,
}) => {
  const linearGradientID = "linGrad";
  const clipGradientID = "clipGrad";
  const buttonColor = getColors(theme);
  const { width: staticWidth, height } = getSize(size);
  const [dynamicWidth, setDynamicWidth] = useState<number | undefined>(
    undefined,
  );
  const buttonWidth = staticWidth ?? dynamicWidth;
  const isAutoSize = size === "auto";

  const handleLayout = useCallback(
    (event: LayoutChangeEvent) => {
      if (isAutoSize) {
        setDynamicWidth(
          event.nativeEvent.layout.width + (buttonIcon ? 60 : 40),
        );
      }
    },
    [isAutoSize, buttonIcon, setDynamicWidth],
  );

  return (
    <Pressable
      style={StyleSheet.flatten([
        s.container,
        {
          marginTop,
          marginBottom,
          marginLeft,
          marginRight,
          height,
          width: buttonWidth,
          alignSelf: isAutoSize ? "flex-start" : "auto",
        },
      ])}
      onPress={onPress}
      testID="button-pressable-id"
    >
      {!!buttonWidth && (
        <Svg
          width={buttonWidth}
          height={height}
          style={StyleSheet.absoluteFillObject}
        >
          <Defs>
            <LinearGradient
              id={linearGradientID}
              x1="0%"
              x2="100%"
              y1="0%"
              y2="100%"
            >
              <Stop offset="0%" stopColor={buttonColor[0]} />
              <Stop offset="100%" stopColor={buttonColor[1]} />
            </LinearGradient>
            <ClipPath id={clipGradientID}>
              <Rect
                x="0"
                y="0"
                width={buttonWidth}
                height={height}
                rx="25"
                ry="50"
              />
            </ClipPath>
          </Defs>
          <Rect
            x="0"
            y="0%"
            width={buttonWidth}
            height={height}
            clipPath={`url(#${clipGradientID})`}
            fill={`url(#${linearGradientID})`}
          />
        </Svg>
      )}
      <View style={s.contentContainer} onLayout={handleLayout}>
        {!!buttonIcon && buttonIcon()}
        <Text style={s.defaultText}>{text}</Text>
      </View>
    </Pressable>
  );
};

export default Button;
