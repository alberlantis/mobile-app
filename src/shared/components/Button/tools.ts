import { DimensionValue, ColorValue, StyleProp, ViewStyle } from "react-native";

import colors from "src/theme/colors";

type NumberProp = string | number;
export type ButtonSize = "fill" | "auto";
export type ButtonTheme =
  | "primary"
  | "secondary"
  | "disabled"
  | "off"
  | "primary-outline";
export type ButtonDimensions = {
  width: number;
  height: number;
};

export const getColors = (theme: ButtonTheme): ColorValue[] => {
  switch (theme) {
    case "secondary":
      return [colors.BLUE_PRIMARY, colors.BLUE_PRIMARY];
    case "disabled":
      return [colors.GRAY_BOLD, colors.GRAY_BOLD];
    case "off":
      return [colors.WHITE_GRAY, colors.WHITE_GRAY];
    case "primary-outline":
      return [colors.TRANSPARENT, colors.TRANSPARENT];
    case "primary":
    default:
      return [colors.ORANGE_PRIMARY_DARK, colors.ORANGE_PRIMARY_LIGHT];
  }
};

export const getSize = (size?: ButtonSize): DimensionValue => {
  switch (size) {
    case "fill": {
      return "100%";
    }
    default:
      return "auto";
  }
};

export const getRadius = (dynamicSize: ButtonDimensions): NumberProp => {
  const buttonRadiusRation = 200 / 300;
  return buttonRadiusRation * Math.min(dynamicSize.width, dynamicSize.height);
};

export const getOutline = (
  theme: ButtonTheme,
  radius: NumberProp,
): StyleProp<ViewStyle> => {
  if (!theme.includes("outline")) return {};
  switch (theme) {
    case "primary-outline":
    default: {
      return {
        borderRadius: radius as number,
        borderColor: colors.ORANGE_PRIMARY,
        borderWidth: 1,
      };
    }
  }
};
