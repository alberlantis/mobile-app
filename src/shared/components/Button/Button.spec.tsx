import { fireEvent, render, screen } from "@testing-library/react-native";
import { StyleSheet } from "react-native";

import s from "./Button.style";
import Button, { type IButtonProps } from "./Button";

const renderComponent = (props: IButtonProps) => {
  render(<Button {...props} />);
};

const mockOnPress = jest.fn();

describe("Button", () => {
  it("should render primary large button", () => {
    renderComponent({
      theme: "primary",
      size: "large",
      text: "primary large",
      onPress: mockOnPress,
    });

    const button = screen.getByTestId("button-pressable-id");
    expect(screen.toJSON()).toMatchSnapshot();
    expect(screen.getByText("primary large")).toBeOnTheScreen();
    expect(button).toBeOnTheScreen();
    expect(button).toHaveStyle(
      StyleSheet.flatten([
        s.container,
        { width: 300, height: 50 },
        { marginTop: undefined, marginBottom: undefined },
      ]),
    );
  });

  it("should render primary extra-large button", () => {
    renderComponent({
      theme: "primary",
      size: "extra-large",
      text: "primary large",
      onPress: mockOnPress,
    });

    const button = screen.getByTestId("button-pressable-id");
    expect(screen.toJSON()).toMatchSnapshot();
    expect(screen.getByText("primary large")).toBeOnTheScreen();
    expect(button).toBeOnTheScreen();
    expect(button).toHaveStyle(
      StyleSheet.flatten([
        s.container,
        { width: "100%", height: 50 },
        { marginTop: undefined, marginBottom: undefined },
      ]),
    );
  });

  it("should render secondary large button", () => {
    renderComponent({
      theme: "secondary",
      size: "large",
      text: "secondary large",
      onPress: mockOnPress,
    });

    const button = screen.getByTestId("button-pressable-id");
    expect(screen.toJSON()).toMatchSnapshot();
    expect(screen.getByText("secondary large")).toBeOnTheScreen();
    expect(button).toBeOnTheScreen();
    expect(button).toHaveStyle(
      StyleSheet.flatten([
        s.container,
        { width: 300, height: 50 },
        { marginTop: undefined, marginBottom: undefined },
      ]),
    );
  });

  it("should render secondary extra-large button", () => {
    renderComponent({
      theme: "secondary",
      size: "extra-large",
      text: "secondary large",
      onPress: mockOnPress,
    });

    const button = screen.getByTestId("button-pressable-id");
    expect(screen.toJSON()).toMatchSnapshot();
    expect(screen.getByText("secondary large")).toBeOnTheScreen();
    expect(button).toBeOnTheScreen();
    expect(button).toHaveStyle(
      StyleSheet.flatten([
        s.container,
        { width: "100%", height: 50 },
        { marginTop: undefined, marginBottom: undefined },
      ]),
    );
  });

  it("should trigger onPress when is tapped", () => {
    renderComponent({
      theme: "primary",
      size: "large",
      text: "secondary large",
      onPress: mockOnPress,
    });
    const button = screen.getByTestId("button-pressable-id");
    fireEvent.press(button);
    expect(mockOnPress).toHaveBeenCalledTimes(1);
  });
});
