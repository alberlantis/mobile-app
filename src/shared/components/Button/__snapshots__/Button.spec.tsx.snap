// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render primary extra-large button 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "height": 50,
      "justifyContent": "center",
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "width": "100%",
    }
  }
  testID="button-pressable-id"
>
  <RNSVGSvgView
    bbHeight={50}
    bbWidth="100%"
    focusable={false}
    height={50}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "flex": 0,
          "height": 50,
          "width": "100%",
        },
      ]
    }
    width="100%"
  >
    <RNSVGGroup
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
    >
      <RNSVGDefs>
        <RNSVGLinearGradient
          gradient={
            [
              0,
              -34743,
              1,
              -29381,
            ]
          }
          gradientTransform={null}
          gradientUnits={0}
          name="linGrad"
          x1="0%"
          x2="100%"
          y1="0%"
          y2="100%"
        />
        <RNSVGClipPath
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          name="clipGrad"
        >
          <RNSVGRect
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            height={50}
            rx="25"
            ry="50"
            width="100%"
            x="0"
            y="0"
          />
        </RNSVGClipPath>
      </RNSVGDefs>
      <RNSVGRect
        clipPath="clipGrad"
        fill={
          {
            "brushRef": "linGrad",
            "type": 1,
          }
        }
        height={50}
        propList={
          [
            "fill",
          ]
        }
        width="100%"
        x="0"
        y="0%"
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    onLayout={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        {
          "color": "#FFF",
          "fontSize": 16,
          "fontWeight": "medium",
        }
      }
    >
      primary large
    </Text>
  </View>
</View>
`;

exports[`Button should render primary large button 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "height": 50,
      "justifyContent": "center",
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "width": 300,
    }
  }
  testID="button-pressable-id"
>
  <RNSVGSvgView
    bbHeight={50}
    bbWidth={300}
    focusable={false}
    height={50}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "flex": 0,
          "height": 50,
          "width": 300,
        },
      ]
    }
    width={300}
  >
    <RNSVGGroup
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
    >
      <RNSVGDefs>
        <RNSVGLinearGradient
          gradient={
            [
              0,
              -34743,
              1,
              -29381,
            ]
          }
          gradientTransform={null}
          gradientUnits={0}
          name="linGrad"
          x1="0%"
          x2="100%"
          y1="0%"
          y2="100%"
        />
        <RNSVGClipPath
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          name="clipGrad"
        >
          <RNSVGRect
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            height={50}
            rx="25"
            ry="50"
            width={300}
            x="0"
            y="0"
          />
        </RNSVGClipPath>
      </RNSVGDefs>
      <RNSVGRect
        clipPath="clipGrad"
        fill={
          {
            "brushRef": "linGrad",
            "type": 1,
          }
        }
        height={50}
        propList={
          [
            "fill",
          ]
        }
        width={300}
        x="0"
        y="0%"
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    onLayout={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        {
          "color": "#FFF",
          "fontSize": 16,
          "fontWeight": "medium",
        }
      }
    >
      primary large
    </Text>
  </View>
</View>
`;

exports[`Button should render secondary extra-large button 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "height": 50,
      "justifyContent": "center",
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "width": "100%",
    }
  }
  testID="button-pressable-id"
>
  <RNSVGSvgView
    bbHeight={50}
    bbWidth="100%"
    focusable={false}
    height={50}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "flex": 0,
          "height": 50,
          "width": "100%",
        },
      ]
    }
    width="100%"
  >
    <RNSVGGroup
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
    >
      <RNSVGDefs>
        <RNSVGLinearGradient
          gradient={
            [
              0,
              -11245864,
              1,
              -11245864,
            ]
          }
          gradientTransform={null}
          gradientUnits={0}
          name="linGrad"
          x1="0%"
          x2="100%"
          y1="0%"
          y2="100%"
        />
        <RNSVGClipPath
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          name="clipGrad"
        >
          <RNSVGRect
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            height={50}
            rx="25"
            ry="50"
            width="100%"
            x="0"
            y="0"
          />
        </RNSVGClipPath>
      </RNSVGDefs>
      <RNSVGRect
        clipPath="clipGrad"
        fill={
          {
            "brushRef": "linGrad",
            "type": 1,
          }
        }
        height={50}
        propList={
          [
            "fill",
          ]
        }
        width="100%"
        x="0"
        y="0%"
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    onLayout={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        {
          "color": "#FFF",
          "fontSize": 16,
          "fontWeight": "medium",
        }
      }
    >
      secondary large
    </Text>
  </View>
</View>
`;

exports[`Button should render secondary large button 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "height": 50,
      "justifyContent": "center",
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "width": 300,
    }
  }
  testID="button-pressable-id"
>
  <RNSVGSvgView
    bbHeight={50}
    bbWidth={300}
    focusable={false}
    height={50}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "flex": 0,
          "height": 50,
          "width": 300,
        },
      ]
    }
    width={300}
  >
    <RNSVGGroup
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
    >
      <RNSVGDefs>
        <RNSVGLinearGradient
          gradient={
            [
              0,
              -11245864,
              1,
              -11245864,
            ]
          }
          gradientTransform={null}
          gradientUnits={0}
          name="linGrad"
          x1="0%"
          x2="100%"
          y1="0%"
          y2="100%"
        />
        <RNSVGClipPath
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          name="clipGrad"
        >
          <RNSVGRect
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            height={50}
            rx="25"
            ry="50"
            width={300}
            x="0"
            y="0"
          />
        </RNSVGClipPath>
      </RNSVGDefs>
      <RNSVGRect
        clipPath="clipGrad"
        fill={
          {
            "brushRef": "linGrad",
            "type": 1,
          }
        }
        height={50}
        propList={
          [
            "fill",
          ]
        }
        width={300}
        x="0"
        y="0%"
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    onLayout={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        {
          "color": "#FFF",
          "fontSize": 16,
          "fontWeight": "medium",
        }
      }
    >
      secondary large
    </Text>
  </View>
</View>
`;
